### PR TITLE
Refactored datastorage mock logging

### PR DESCRIPTION
**Pull request by hAdamm**

### CODE REVIEW

File: tests/__assets/DataStorage/DataStorage.php [hunk 1]
```diff
@@ -4,39 +4,55 @@
...
-    const EVENT_TYPE_INSERT = "inserts";
-    const EVENT_TYPE_UPDATE = "updates";
-    const EVENT_TYPE_DELETE = "deletes";
+    const EVENT_TYPE_INSERT = "insert";
+    const EVENT_TYPE_UPDATE = "update";
+    const EVENT_TYPE_DELETE = "delete";
+    const EVENT_TYPE_CONSTRUCT_QUERY = "constructQuery";
+    const EVENT_TYPE_COUNT_ENTITIES_IN_RELATIONSHIP = "countEntitiesInRelationship";
+    const EVENT_TYPE_COUNT_RECORDS_FOR_QUERY = "countRecordsForQuery";
+    const EVENT_TYPE_CREARE_RELATIONSHIP_BETWEEN_ENTITIES = "createRelationshipBetweenEntities";
```
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: create

File: tests/__assets/DataStorage/DataStorage.php [hunk 2]
```diff
@@ -4,39 +4,55 @@
...
 
     public static $autoIncrementIndex = 0;
 
     protected $eventsStack = [
-            "inserts" => ["all" => [], "byEntities" => []],
-            "updates" => ["all" => [], "byEntities" => []],
-            "deletes" => ["all" => [], "byEntities" => []],
+            "insert" => ["all" => [], "byKeys" => []],
+            "update" => ["all" => [], "byKeys" => []],
+            "delete" => ["all" => [], "byKeys" => []],
```
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: why u are not using those constants here?

File: tests/__assets/DataStorage/DataStorage.php [hunk 3]
```diff
@@ -4,39 +4,55 @@
...
-    public function getAllEventsForTypeAndEntity($type, \obo\Entity $entity) {
-        return $this->eventsStack[$type]["byEntities"][$entity->objectIdentificationKey()];
+    public function getAllEventsForTypeAndKey($type, $key) {
+        return isset($this->eventsStack[$type]["byKeys"][$key]) ? $this->eventsStack[$type]["byKeys"][$key] : [];
     }
 
-    public function getLastEventForTypeAndEntity($type, \obo\Entity $entity) {
-        return \end($this->eventsStack[$type]["byEntities"][$entity->objectIdentificationKey()]);
+    public function getLastEventForTypeAndKey($type, $key) {
+        return isset($this->eventsStack[$type]["byKeys"][$key]) ? \end($this->eventsStack[$type]["byKeys"][$key]) : false;
```
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: return values are not consistent (at least I think) in all 4 methods

File: tests/__assets/DataStorage/DataStorage.php [hunk 4]
```diff
@@ -49,95 +65,142 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
                 function ($repositoryName, $entities) {
                     return (new static)->createRelationshipBetweenEntities($repositoryName, $entities);
                 }
             );
     }
 
-    public static function setDefaultDataForEntitiesInRelationshipBehavior($dataStorageMock) {
+    public function dataForEntitiesInRelationship(\obo\Carriers\QueryCarrier $specification, $repositoryName, \obo\Entity $owner, $targetEntity) {
+        $this->logEventForDataForEntitiesInRelationship($specification, $repositoryName, $owner, $targetEntity);
+        return [];
```
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: why?

File: tests/__assets/DataStorage/DataStorage.php [hunk 5]
```diff
@@ -49,95 +65,142 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function setDefaultDataForEntitiesInRelationshipBehavior($dataStorageMock) {
         $dataStorageMock->shouldReceive("dataForEntitiesInRelationship")->andReturnUsing(
                 function ($specification, $repositoryName, $owner, $targetEntity) {
                     return (new static)->dataForEntitiesInRelationship($specification, $repositoryName, $owner, $targetEntity);
                 }
             );
     }
 
+    public function dataForQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
+        //throw new Exception();
```
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: :)

File: tests/__assets/DataStorage/DataStorage.php [hunk 6]
```diff
@@ -49,95 +65,142 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+        $this->logEventForDataForQuery($queryCarrier);
+        return [];
```
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: again

File: tests/__assets/DataStorage/DataStorage.php [hunk 7]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForUpdateEntity(\obo\Entity $entity) {
+        $this->logEvent(["dataToStore" => $entity->dataToStore(), "entity" => $entity], static::EVENT_TYPE_UPDATE, $entity->objectIdentificationKey());
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableUpdateEntityLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('updateEntity')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 8]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForRepositoryAddressForEntity(\obo\Entity $entity) {
+        $this->logEvent(["entity" => $entity], static::EVENT_TYPE_REPOSITORY_ADDRESS_FOR_ENTITY, $entity->objectIdentificationKey());
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableRepositoryAddressForEntityLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('repositoryAddressForEntity')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 9]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForRemoveRelationshipBetweenEntities($repositoryName, array $entities) {
+        $this->logEvent(["repositoryName" => $repositoryName, "entities" => $entities], static::EVENT_TYPE_REMOVE_RELATIONSHIP_BETWEEN_ENTITIES);
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableRemoveRelationshipBetweenEntitiesLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('removeRelationshipBetweenEntities')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 10]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForRemoveEntity(\obo\Entity $entity) {
+        $this->logEvent([$entity->entityInformation()->primaryPropertyName => $entity->primaryPropertyValue(), "softDelete" => $entity->entityInformation()->propertyNameForSoftDelete, "entity" => $entity], static::EVENT_TYPE_DELETE, $entity->objectIdentificationKey());
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableRemoveEntityLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('removeEntity')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 11]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForInsertEntity(\obo\Entity $entity) {
+        $this->logEvent(["dataToStore" => $entity->dataToStore(), "entity" => $entity], static::EVENT_TYPE_INSERT, $entity->objectIdentificationKey());
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableInsertEntityLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('insertEntity')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 12]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
                 function ($specification, $repositoryName, $owner, $targetEntity) {
                     return (new static)->dataForEntitiesInRelationship($specification, $repositoryName, $owner, $targetEntity);
                 }
             );
     }
 
-    public function setDefaultDataForQueryBehavior($dataStorageMock) {
+    /**
+     * @param \obo\Carriers\QueryCarrier $queryCarrier
+     * @return arrya
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: arraya :dagger:

File: tests/__assets/DataStorage/DataStorage.php [hunk 13]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForDataForQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
+        $this->logEvent(["queryCarrier" => $queryCarrier], static::EVENT_TYPE_DATA_FOR_QUERY, $queryCarrier->getDefaultEntityClassName());
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableDataForQueryLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('dataForQuery')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 14]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForDataForEntitiesInRelationship(\obo\Carriers\QueryCarrier $specification, $repositoryName, \obo\Entity $owner, $targetEntity) {
+        $this->logEvent(["queryCarrier" => $specification, "repositoryName" => $repositoryName, "owner" => $owner, "targetEntity" => $targetEntity], static::EVENT_TYPE_DATA_FOR_ENTITIES_IN_RELATIONSHIP, $specification->getDefaultEntityClassName());
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableDataForEntitiesInRelationshipLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('dataForEntitiesInRelationship')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 15]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
+    public function logEventForCreateRelationshipBetweenEntities($repositoryName, array $entities) {
+        $this->logEvent(["repositoryName" => $repositoryName, "entities" => $entities], static::EVENT_TYPE_CREATE_RELATIONSHIP_BETWEEN_ENTITIES);
+    }
+
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableCreateRelationshipBetweenEntitiesLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('createRelationshipBetweenEntities')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 16]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
 
-    public function updateEntity(\obo\Entity $entity) {
-        $this->logEventForEntity($entity->dataToStore(), static::EVENT_TYPE_UPDATE, $entity);
-        return null;
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableCountRecordsForQueryLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('countRecordsForQuery')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 17]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
-        $this->logEventForEntity($entity->dataToStore(), static::EVENT_TYPE_INSERT, $entity);
-        if ($entity->entityInformation()->informationForPropertyWithName($entity->entityInformation()->primaryPropertyName)->autoIncrement) {
-            $entity->setValueForPropertyWithName(++static::$autoIncrementIndex, $entity->entityInformation()->primaryPropertyName, false);
-        }
+    /**
+     * @param type $dataStorageMock
+     * @return void
+     */
+    public function enableCountEntitiesInRelationshipLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('countEntitiesInRelationship')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php [hunk 18]
```diff
@@ -49,136 +93,450 @@ public function constructQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
...
     }
 
-    public function countRecordsForQuery(\obo\Carriers\QueryCarrier $queryCarrier) {
-        return 0;
+    /**
+     * @param \Mockery\MockInterface $dataStorageMock
+     * @return void
+     */
+    public function enableConstructQueryLogging(\Mockery\MockInterface $dataStorageMock) {
+        $dataStorageMock->shouldReceive('constructQuery')
```
Comment (ZZromanZZ) [inline tests/__assets/DataStorage/DataStorage.php]: single quotes

File: tests/__assets/DataStorage/DataStorage.php (hunk 12 above)
Comment (lordjim666) [inline tests/__assets/DataStorage/DataStorage.php]: lol! https://www.youtube.com/watch?v=iN8PKcNGcuI

**Pull request merged**